src/extension.ts

```diff
@@ -1,206 +1,345 @@
       })
-    )
+    })
 
     // A helper command, chains up to other commands based on current pipeline state
-    context.subscriptions.push(
-      registerCommand(`${EXT_ID}.build`, async () => {
-        runner.completeBuild = true
-        if (!store.state.getState().pipeline.initialized) {
-          await executeCommand(`${EXT_ID}.${TaskMode.buildInit}`)
-        } else if (!store.state.getState().pipeline.dependencies.updated) {
-          await executeCommand(`${EXT_ID}.${TaskMode.updateDeps}`)
-        } else if (!store.state.getState().pipeline.dependencies.built) {
-          await executeCommand(`${EXT_ID}.${TaskMode.buildDeps}`)
-        } else if (!store.state.getState().pipeline.application.built) {
-          await executeCommand(`${EXT_ID}.${TaskMode.buildApp}`)
-        } else {
-          terminal.appendMessage('Nothing to do')
-        }
-      })
+    this.registerCommand('build', async () => {
+      this.runner.completeBuild = true
+      if (!this.workspaceState.getInitialized()) {
+        await this.executeCommand(TaskMode.buildInit)
+      } else if (!this.workspaceState.getDependenciesUpdated()) {
+        await this.executeCommand(TaskMode.updateDeps)
+      } else if (!this.workspaceState.getDependenciesBuilt()) {
+        await this.executeCommand(TaskMode.buildDeps)
+      } else if (!this.workspaceState.getApplicationBuilt()) {
+        await this.executeCommand(TaskMode.buildApp)
+      } else {
+        this.outputTerminal.appendMessage('Nothing to do')
+      }
+    })
+  }
+
+  async deactivate() {
+    const activeManifest = this.manifestManager.getActiveManifest()
+    if (activeManifest) {
+      await this.deactivateIntegrations(activeManifest)
+    }
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  private registerCommand(name: string, callback: (...args: any) => any | Promise<void>) {
+    this.extCtx.subscriptions.push(
+      commands.registerCommand(`${EXT_ID}.${name}`, callback)
     )
   }
-}
 
-export async function deactivate(): Promise<void> {
-  const manifest = store.state.getState().selectedManifest
-  if (manifest) {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  private executeCommand(name: string, ...args: any[]): Thenable<unknown> {
+    return commands.executeCommand(`${EXT_ID}.${name}`, args)
+  }
+
+  private async handleActiveManifestChanged(manifest: FlatpakManifest | null, isLastActive: boolean) {
+    if (manifest === null) {
+      return
+    }
+
+    if (!isLastActive) {
+      this.runner.close()
+      await this.resetPipelineState()
+
+      this.terminalProvider?.dispose()
+      this.terminalProvider = undefined
+
+      for (const terminal of this.activeTerminals) {
+        terminal.dispose()
+      }
+
+      await manifest.deleteRepoDir()
+    }
+
+    this.terminalProvider = new TerminalProvider(manifest)
+
+    await this.ensurePipelineState()
+  }
+
+  private async resetPipelineState(): Promise<void> {
```

Comment:
Should probably be part of workspaceState and called reset :)

Reply:
we dont want to reset activeManifestUri too. I removed it in the workspaceState as workspaceState shoudn't know which keys are part of the pipeline state